### PR TITLE
Models 'Forwarded' header

### LINKED ISSUE #3471: Support parsing the 'Forwarded' header: RFC 7239
There's a specification [1,2] for a header which does not have native parsing support in http4s. The `Forwarded` header can be included by proxies as a standards-track alternative to `X-Forwarded-For`, `X-Forwarded-Proto`, and `X-Forwarded-Host`. http4s should consider providing support for it.

[1] [Forwarded - HTTP | MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Forwarded)
[2] [RFC 7239 - Forwarded HTTP Extension](https://tools.ietf.org/html/rfc7239])

### PR DESCRIPTION
**Pull request by satorg**
Resolves #3471.

The `Forwarded` header model tries to be as type safe as possible and thus introduces a hierarchy of model classes for that.

In particular, `Element` enables the following syntax for building elements:
```scala
Element
  .fromFor(Node(...))
  .withBy(Node(...))
  .withHost(Host(...))
  .withProto(Proto(...))
```

### CODE REVIEW

**Comment by hamnis** [comment]
Great stuff. =)

This should probably be targeted 0.21 ?

**Comment by satorg** [comment]
I guess it could... Would you suggest to rebase it onto `series/0.21` ?

**Comment by hamnis** [comment]
@satorg Yes

**Comment by hamnis** [comment]
We should have some tests for this, I do believe there are some examples in the RFC

**Comment by satorg** [comment]
Yes, for sure. I am working on the tests now, per se.

File: core/src/main/scala/org/http4s/headers/Forwarded.scala
Comment (rossabaker) [inline core/src/main/scala/org/http4s/headers/Forwarded.scala:26]: I've never been good at predicting `AnyVal` boxing, but since it's polymorphic in `Node`, won't these always be boxed in practice?  I'm not sure the `Any` / `AnyRef` is buying anything here, but I could be wrong.

File: core/src/main/scala/org/http4s/headers/Forwarded.scala
Comment (rossabaker) [inline core/src/main/scala/org/http4s/headers/Forwarded.scala:55]: Not every string is a valid `obfnode` or `obfport`.  This should probably have a private constructor and a `fromString` to prevent invalid states.

Comment (rossabaker) [inline core/src/main/scala/org/http4s/headers/Forwarded.scala:55]: Reading the spec, I expected `obfnode` to be in `Node` and `obfport` to be in `Port`.  The grammar is the same for this type, but it seems weird to me for the type to do double duty.

File: core/src/main/scala/org/http4s/headers/Forwarded.scala
Comment (rossabaker) [inline core/src/main/scala/org/http4s/headers/Forwarded.scala:46]: If we're being real strict about the spec, only values 0 to 99999 are allowed.  (Sadly, not 0 to 65535, or we could wrap a `Short`!)

File: core/src/main/scala/org/http4s/headers/Forwarded.scala
Comment (rossabaker) [inline core/src/main/scala/org/http4s/headers/Forwarded.scala:79]: We use `withFoo` instead of `foo=` elsewhere.  Maybe that's better here?

File: core/src/main/scala/org/http4s/headers/Forwarded.scala
Comment (rossabaker) [inline core/src/main/scala/org/http4s/headers/Forwarded.scala:112]: I've grown regretful over a lot of our type aliases over the years.  Is anything motivating this one?

File: core/src/main/scala/org/http4s/parser/ForwardedHeader.scala
Comment (rossabaker) [inline core/src/main/scala/org/http4s/parser/ForwardedHeader.scala:34]: I remember a conversation about this.  Are we tracking this somewhere?

File: core/src/main/scala/org/http4s/headers/Forwarded.scala
Comment (satorg) [reply core/src/main/scala/org/http4s/headers/Forwarded.scala:55]: Agree. There's still plenty of stuff to do.

Comment (satorg) [reply core/src/main/scala/org/http4s/headers/Forwarded.scala:55]: I agree that it is a controversial approach. I was just thinking – what would be advantages of making them separated?
I mean it would not only require to create two different (although identical) case classes here, but also two parser rules for creating these classes. No doubt we could share some of their implementation, but it would increase overall code complexity anyway (remember, there still should be `fromString` as well as rendering implemented for both of them). On the other side, I couldn't come up with some real disadvantages besides, yeah, some weirdness.

I'm just trying to come up with some compromise solution that would let us avoid code duplication, preserve type safety and keep code complexity on some maintainable level. WDYT?

File: core/src/main/scala/org/http4s/headers/Forwarded.scala
Comment (satorg) [reply core/src/main/scala/org/http4s/headers/Forwarded.scala:46]: Well, I could make its ctor private and add something like `fromInt` method to `Port` if you think it might make sense.

Comment (hamnis) [reply core/src/main/scala/org/http4s/headers/Forwarded.scala:46]: `fromInt(port: Int): Either[ParseError, Port]` might make sense

File: core/src/main/scala/org/http4s/headers/Forwarded.scala
Comment (satorg) [reply core/src/main/scala/org/http4s/headers/Forwarded.scala:79]: Yeah, I anticipated you won't like it:)
Actually, I was just playing a bit with different approaches.
I agree – it's better to keep this DSL consistent with others parts.

File: core/src/main/scala/org/http4s/headers/Forwarded.scala
Comment (satorg) [reply core/src/main/scala/org/http4s/headers/Forwarded.scala:112]: Just traces of my experiments with the model... I'll remove it.

File: core/src/main/scala/org/http4s/headers/Forwarded.scala
Comment (satorg) [reply core/src/main/scala/org/http4s/headers/Forwarded.scala:26]: I added these at the very beginning of my work, just in case, but I agree – there's no real sense to keep them.

File: core/src/main/scala/org/http4s/parser/ForwardedHeader.scala
Comment (satorg) [reply core/src/main/scala/org/http4s/parser/ForwardedHeader.scala:34]: I guess nope.
The original question I raised is here: https://gitter.im/http4s/http4s-dev?at=5f056255dbf01050ab5e7a14.
I'll create a separate issue for that, if you don't mind.

**Comment by satorg** [comment]
> I was surprised how complicated this is, until I read the spec. This is really nice work.

Yeah, I also didn't realize that until I dove into this task. Unfortunately, due to my personal circumstances, I just didn't have enough time to squeeze it into 0.21.7, I am sorry for that.

File: core/src/main/scala/org/http4s/headers/Forwarded.scala
Comment (rossabaker) [reply core/src/main/scala/org/http4s/headers/Forwarded.scala:55]: This comment is about minutiae, so I don't feel too strongly.

`Obfuscated` isn't as untyped as `String`, but it still raises the question, "an obfuscated what?"  It's not a particularly meaningful type on its own.  Sharing makes the implementation more concise, but isn't semantically meaningful in an HTTP sense.

In the unlikely event that the next version of that RFC allows some other character in an obfuscated host, we've got room to change it.  If the two unrelated concepts share a type, we'd have to wait for the next breaking release to implement it.

I don't think either argument matters much in this corner case of a supplemental RFC.  I'm thinking more in general design.  And you're right that the repetition is annoying, so I'm content either way.

File: core/src/main/scala/org/http4s/headers/Forwarded.scala
Comment (rossabaker) [reply core/src/main/scala/org/http4s/headers/Forwarded.scala:79]: It reminded me of Scala's `_=` operator syntax, but those are for side effects.  If your convention didn't require the backticks, I'd probably like it a lot! :smile:

File: core/src/main/scala/org/http4s/parser/ForwardedHeader.scala
Comment (rossabaker) [reply core/src/main/scala/org/http4s/parser/ForwardedHeader.scala:34]: That'd be great.  Sorry I let that one fall through the cracks.

**Comment by rossabaker** [comment]
No apologies required at all.  Glad for the help, and we won't run out of patch numbers any time soon.

File: core/src/main/scala/org/http4s/headers/Forwarded.scala
Comment (satorg) [reply core/src/main/scala/org/http4s/headers/Forwarded.scala]: @rossabaker I wonder how come the port range you mentioned is 0..99999? I looked through a bunch of specs but it seems that everywhere it is 0..65535, even for IPv6.

Comment (satorg) [reply core/src/main/scala/org/http4s/headers/Forwarded.scala]: @hamnis if we would like to guarantee correctness of the port number in `Forwarded.Node.Port` instances, then perhaps it makes sense to do the same for `Forwarded.Host`. But I'm struggling to come up with some nice name for such constructor, because it takes two parameters – `Uri.Host` and `Int` simultaneously. Would something like
```
object Host {
  def parse(uriHost: Uri.Host, portNum: Int): ParseResult[Host]
}
```
be good enough or could you suggest some nicer naming?

Comment (hamnis) [reply core/src/main/scala/org/http4s/headers/Forwarded.scala]: `parse` does not really sit well with me here.
what about just `from` 

maybe a constructor from `Uri` might make sense as well.

`def fromUri(uri: Uri): ParseResult[Host]`

File: core/src/main/scala/org/http4s/headers/Forwarded.scala
Comment (hamnis) [inline core/src/main/scala/org/http4s/headers/Forwarded.scala:110]: this comment is not correct anymore

Comment (satorg) [reply core/src/main/scala/org/http4s/headers/Forwarded.scala:110]: Good catch, thanks!

File: core/src/main/scala/org/http4s/headers/Forwarded.scala
Comment (satorg) [inline core/src/main/scala/org/http4s/headers/Forwarded.scala:119]: Back-ticks are here only because of `for` – apparently we cannot use it as an identifier.
All other field names don't need back-ticks (apparently), but I found it would be ugly to back-tick `for` only. But it looks quite ugly anyway. So I'd appreciate any better naming suggestions here:)

File: core/src/main/scala/org/http4s/parser/ForwardedHeader.scala
Comment (satorg) [reply core/src/main/scala/org/http4s/parser/ForwardedHeader.scala:34]: Created the issue #3657 (for reference)

**Comment by satorg** [comment]
Ok, it seems I'm getting close. I have implemented the model renderers and mostly need to add some tests for them.

**Comment by satorg** [comment]
Looks like test failures are due to #3664...

**Comment by jyoo980** [comment]
@satorg sorry for the hold up, I've written up my investigations into the flaky failures in #3664

**Comment by satorg** [comment]
@jyoo980 no worries, I'm still working on my tests here anyway.

**Comment by satorg** [comment]
I believe I'm pretty much done with this PR.
Hopefully, it reproduces the corresponding RFC as close as possible (with an exception of the controversial approach of sharing the `Obfuscated` type between `Node.Name` and `Node.Port`).
Anyway I'm open for any suggestions, feel free to point them out.

**cc** @rossabaker, @hamnis

File: tests/src/test/scala/org/http4s/headers/ForwardedArbitraryInstances.scala
Comment (satorg) [inline tests/src/test/scala/org/http4s/headers/ForwardedArbitraryInstances.scala:18]: I wonder, should this trait be moved to `org.http4s.laws.discipline` (where `ArbitraryInstances` sits)?

**Review by hamnis — APPROVED** [review]

**Comment by hamnis** [comment]
A colleage of mine just hit this in production https://github.com/scala/bug/issues/8831 
Could this be a problem with the backticks for the parameters ?

**Comment by satorg** [comment]
@hamnis, interesting issue, never seen it before. Thanks for pointing this out to me!

I'd guess it's unlikely to affect this code since the bug usually appears when there are some common prefixes among parameters.

But to be on the safer side, I agree, we can consider getting rid of backticks. As I wrote some time before, the backticks are only necessary for the `for` parameter, while the other parameters get backticked just to look consistently with `for`.

But it is not the only option – we could rename all the parameters somehow else. For example, since all the parameters are actually optional, they could be renamed to
```scala
sealed trait Element extends Renderable { _: Product =>
    def maybeBy: Option[Node]
    def maybeFor: Option[Node]
    def maybeHost: Option[Host]
    def maybeProto: Option[Proto]
...
}
```
..or something like that. WDYT?

**Comment by hamnis** [comment]
I've never been a fan of the backticks of for instance header classes just to get the "correct" name, it makes them less discoverable by for instance IDEA. So I prefer the latter version here. Other maintainers may prefer differently. 

@rossabaker any guidelines here?

**Comment by satorg** [comment]
Looks like the most recent scalafmt has broken this PR. I'll take care of it.

@rossabaker, btw, I'd appreciate your thoughts regarding the questions above. Feel free to make any suggestions please – I'm open to polish the PR as much as possible.

File: core/src/main/scala/org/http4s/parser/ForwardedHeader.scala
Comment (satorg) [inline core/src/main/scala/org/http4s/parser/ForwardedHeader.scala:36]: If by chance we get #3743 merged before this one, it may make sense to remove this override.

File: tests/src/test/scala/org/http4s/headers/ForwardedArbitraryInstances.scala
Comment (rossabaker) [reply tests/src/test/scala/org/http4s/headers/ForwardedArbitraryInstances.scala:18]: I think yes, but that doesn't need to block this.

File: core/src/main/scala/org/http4s/headers/Forwarded.scala
Comment (rossabaker) [inline core/src/main/scala/org/http4s/headers/Forwarded.scala:43]: A little late on this, but would we be better off with `from*` constructors instead of overriding `apply`?  I've soured a bit on overrides.  It's a fuzzy case: we usually `apply` if it's total, and `from*` if it's a `ParseResult`.  But we usually don't have so many options to make it total.

I could happily go either way here.

File: core/src/main/scala/org/http4s/headers/Forwarded.scala
Comment (rossabaker) [inline core/src/main/scala/org/http4s/headers/Forwarded.scala:86]: Could avoid an override calling this `fromHost`.

File: core/src/main/scala/org/http4s/headers/Forwarded.scala
Comment (rossabaker) [inline core/src/main/scala/org/http4s/headers/Forwarded.scala:89]: What's the use case for one that takes an `Option`?  This seems most useful if you have a `Uri.Authority`.

File: core/src/main/scala/org/http4s/parser/ForwardedHeader.scala
Comment (rossabaker) [reply core/src/main/scala/org/http4s/parser/ForwardedHeader.scala:36]: Well, it did. :smile:

**Review by rossabaker — COMMENTED** [review]
Thanks for your patience while I dig out of my hole.  This is great work, and everything I mentioned is pretty insignificant.

File: core/src/main/scala/org/http4s/headers/Forwarded.scala
Comment (satorg) [reply core/src/main/scala/org/http4s/headers/Forwarded.scala:43]: Just thoughts... 

First, if we prefer to reduce (or completely get rid of) overrides here, we can safely drop these two constructors:
```
def apply(address: Uri.Ipv4Address): Name
def apply(address: Uri.Ipv6Address): Name
```
Because these two can be replaced with direct calls to `Name.Ipv4` and `Name.Ipv6` – these  `apply`-s are just shortcuts and thus there's no need to add extra `from*` methods for them.

Second, I'd suggest the following names for the remaining constructors:
```
def fromInet4Address(address: Inet4Address) // reflects `Uri.Ipv4Address.fromInet4Address`
def fromIpv4Address(a: Byte, b: Byte, c: Byte, d: Byte) // reflects `Uri.Ipv4Address.apply`

def fromInet6Address(address: Inet6Address): Name // reflects `Uri.Ipv6Address.fromInet6Address`
def fromIpv6Address(a: Short, ..., h: Short): Name // reflects `Uri.Ipv6Address.apply`
```
If you don't mind, I'm up to rename them.

Third, we could consider using different names like `of*` for creating totals when we'd like to avoid `apply` overrides. Just to disambiguate them from `from*`-s, which usually return `ParseResult`. Not sure if it will look consistently across the project, though.

File: core/src/main/scala/org/http4s/headers/Forwarded.scala
Comment (satorg) [reply core/src/main/scala/org/http4s/headers/Forwarded.scala:89]: Well, it is a bit useful in tests where `Gen[Option[Int]]` is used to generate `Host`-s with or without port numbers. But I'm inclining to agree that this override seems redundant.

File: core/src/main/scala/org/http4s/headers/Forwarded.scala
Comment (satorg) [reply core/src/main/scala/org/http4s/headers/Forwarded.scala:86]: Just to clarify: if I remove the next `from` method, then there will be no overrides for `from` anymore. Are you ok to leave this one just as `from` then or still prefer to rename to `fromHost`?

File: core/src/main/scala/org/http4s/headers/Forwarded.scala
Comment (satorg) [reply core/src/main/scala/org/http4s/headers/Forwarded.scala:89]: ... Also it turned out that it is used in parsers :) So some internals leverage this form of `from` a little bit.

File: core/src/main/scala/org/http4s/headers/Forwarded.scala
Comment (rossabaker) [reply core/src/main/scala/org/http4s/headers/Forwarded.scala:43]: :+1: to all of this.  Yeah, I think I like the `of` convention.

File: core/src/main/scala/org/http4s/headers/Forwarded.scala
Comment (rossabaker) [reply core/src/main/scala/org/http4s/headers/Forwarded.scala:86]: Great question.  I think we mostly have `fromTypeOfArgument` elsewhere, and it makes it easier to add more later.

**Comment by satorg** [comment]
@rossabaker @hamnis I have worked through most of your comments, feel free to review the changes once again :)

Moreover, I voluntarily decided to get rid of backticks in `Element`'s field names – I've renamed them to `maybe*` pattern since they're options actually. I've put it into the separate last commit so if you don't like it, we can drop it.

Also please feel free to make any other suggestions, if any. There's no rush on my side, I don't mind to continue polishing if necessary.

**Review by hamnis — APPROVED** [review]

**Pull request merged**
